### PR TITLE
New workshop authoring

### PR DESCRIPTION
**Pull request by merelinguist**
🥇 First pass: ready to merge!

🔋 More advanced features planned for a secondary PR.

Closes: #440 

**Main idea: allow people to draft and publish Markdown workshops without having to touch GitHub**

***

### Features:
- ✨ Custom autosave for multiple drafts—if you leave, your work is preserved
- ⌨️ Full Markdown support, with live two-column preview as you type
- 🏡 Dashboard to manage all your workshops at once
- 🤸‍ Mobile-friendly
- 🧘 More to come!

***

#### Begin
633c78e0-ee0c-4d50-96ed-1557cdbab0fd

#### Create workshop
285fee9e-870c-41f7-935f-1e82648faef0

#### Writing workshops
d4636d0c-1559-4317-a24b-ca365d387df6

#### Reviewing your drafts
bbe25d65-633c-42ed-9c63-4bd7e53d0432

*** 

Paging @lachlanjc

### CODE REVIEW

**Comment by zanedb** [comment]
I love this so much

File: src/pages/workshops/drafts.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  Heading,
+  Text,
+  Icon,
+  theme
+} from '@hackclub/design-system'
+import ReactMarkdown from 'react-markdown'
+import orderBy from 'lodash/orderBy'
+import { Transition } from 'react-spring'
+
+import storage from '../../storage'
```
Comment (lachlanjc) [inline src/pages/workshops/drafts.js]:
```suggestion
import storage from 'storage'
```
We have aliases set up, you can make them all `components/*.js` etc

File: src/pages/workshops/drafts.js [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    font-size: 14px;
+  }
+`
+
+const Left = styled(Box)`
+  max-width: 85%;
+`
+
+const truncate = (str, length) => {
+  const dots = str.length > length ? '...' : ''
```
Comment (lachlanjc) [inline src/pages/workshops/drafts.js]:
```suggestion
  const dots = str.length > length ? '…' : ''
```

File: src/pages/workshops/submit.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+  Icon,
+  Text,
+  IconButton as IconDot,
+  theme,
+  mediaQueries
+} from '@hackclub/design-system'
+import ReactMarkdown from 'react-markdown'
+import Prism from 'prismjs'
+
+import search from '../../search'
```
Comment (lachlanjc) [inline src/pages/workshops/submit.js]:
```suggestion
import search from 'search'
```
Same situation here

File: src/pages/workshops/submit.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  justify: 'center'
+})`
+  min-height: 100vh;
+
+  div {
+    height: fit-content;
+    max-width: 42rem;
+  }
+
+  a {
```
Comment (lachlanjc) [inline src/pages/workshops/submit.js]: `a svg` instead :)

File: src/pages/workshops/submit.js [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+`
+
+export default class extends Component {
+  constructor(props) {
+    super(props)
+    this.state = {
+      view: 'edit',
+      name: search.get('id') || '',
+      value:
+        (storage.get(search.get('id')) && storage.get(search.get('id')).body) ||
```
Comment (lachlanjc) [inline src/pages/workshops/submit.js]: This line is super confusing. Put `search.get('id')` into an `id` variable

**Review by lachlanjc — COMMENTED** [review]
Just a few small code things. I haven't checked out the site yet

**Comment by maxwofford** [comment]
This is awesome, great work @merelinguist. Just have a few UX questions from playing around with it:
1. How do I get to the drafts page? I didn't see a link on the workshops page.
1. What do I do once I have a draft? I don't see a way publish them.

**Comment by merelinguist** [comment]
Thanks, @MaxWofford! To address your questions:

1. I need to add the old Submit button back to the workshops page.

2. Can we add a workshops-specific endpoint to the api? Previously @lachlanjc was using the feedback form for submitting.

**Comment by maxwofford** [comment]
I’d rather just submit through form again. Frankly, we took the old draft page down because we never had a single non-spam submission & I’m hesitant to add more fields to our database until I’m sure we’ll be getting legitimate submissions

**Comment by merelinguist** [comment]
@MaxWofford 👍 Makes sense, will add back the old submission stuff.

**Comment by polytroper** [comment]
screen shot 2019-01-15 at 1 10 32 pm

Layout here is kinda awk. Also why are we getting "(private)" for my email string?

**Comment by merelinguist** [comment]
@lachlanjc @MaxWofford @lachlanjc It works! Happy to take any feedback, but otherwise good to merge. ✨

**Comment by maxwofford** [comment]
cc @polytroper

**Comment by polytroper** [comment]
reviewing

**Comment by lachlanjc** [comment]
Submission works 👍
54bb1284-c166-44b3-849f-cfc3b38c06b0

Incredible work here @merelinguist. This is way better than I’d have done. The design polish here is awesome.

I do think we need to think more about the messaging of encouraging people to submit workshops—as great as this site is, it wasn’t the design holding people back with the old page, it’s that they don’t see the value in investing so much energy to write a workshop. I don’t know what the best solution to that is, but something we should think about.

**Comment by polytroper** [comment]
@lachlanjc Working on that. I think that's going to have to be a
longer-term shift in what exactly a workshop is—before we can start
encouraging people to write more, I think we first need a better handle on
what these are for in the first place.

**Comment by zanedb** [comment]
I just submitted a workshop and no authentication was required — I ignored the login prompts & the "publish" button still worked. Unfortunately, because this just reuses the (anonymous) workshop feedback endpoint, there's no way to force authentication server-side. Maybe just force auth on frontend for now tho? It definitely needs its own endpoint sometime.

**Comment by maxwofford** [comment]
Should users need authentication for submitting a draft? My understanding is this just emails it to staff

Max Wofford
max@maxwofford.com
Engineering & Operations @ Hack Club ( https://hackclub.com )

**Comment by zanedb** [comment]
How else would you get the author of the workshop?

**Comment by polytroper** [comment]
@merelinguist and I touched on this in DMs—we should really hide the drafting page behind the login form. Not sure what the timeline is for implementation.

**Comment by lachlanjc** [comment]
Fabulous work here. I have a short list of edits to make then let’s get this merged today/tomorrow.

**Comment by polytroper** [comment]
:+1:

**Comment by lachlanjc** [comment]
Closing for fixes in #513.

**Pull request closed without merging**